File: python/algo/202608/algo1032.py
from typing import List


class Node:
    __slots__ = 'son', 'end'

    def __init__(self):
        self.son = {}
        self.end = False

class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word:str):  # 插入字符串
        cur = self.root
        for c in word:
            if c not in cur.son:
                cur.son[c] = Node()
            cur = cur.son[c]
        cur.end = True

    def find(self, word:str) -> int:  # 查找字符串
        cur = self.root
        for c in word:
            if c not in cur.son:
                return False
            cur = cur.son[c]
            if cur.end: return True
        return False


class StreamChecker:
    def __init__(self, words: List[str]):
        self.tree = Trie()
        self.stream = ""
        for w in words:
            self.tree.insert(w[::-1])

    def query(self, letter: str) -> bool:
        self.stream += letter
        if len(self.stream) > 200: self.stream = self.stream[1:]
        return self.tree.find(self.stream[::-1])

File: cplus/leetcode/202608/algo1032.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    Node* son[26]{};
    bool end = false;
};

class Trie {
private:
    Node* root =  new Node();
public:
    Trie() {
    }
    
    void insert(string word) {
        auto node = root;
        reverse(word.begin(), word.end());
        for (auto c : word) {
            if (node->son[c-'a'] == nullptr) node->son[c-'a'] = new Node();
            node = node->son[c-'a'];
        }
        node->end = true;
    }
    
    bool find(vector<char>& s) {
        auto node = root;
        for (int i=s.size()-1; i>=0; i--) {
            if (node->son[s[i]-'a'] == nullptr) return false;
            node = node->son[s[i]-'a'];
            if (node->end) return true;
        }

        return false; 
    }
    
};

class StreamChecker {
private:
    vector<char> stream;
    Trie tree;
public:
    StreamChecker(vector<string>& words) {
        for (auto w: words) tree.insert(w);
    }
    
    bool query(char letter) {
        stream.emplace_back(letter);
        return tree.find(stream);
    }
};
